add distributed example for CUDA RPC

gcramer23 opened the pull request:
work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

Differential Revision: [D27420409](https://our.internmc.facebook.com/intern/diff/D27420409)

gcramer23 committed:
test_combining_dpp_with_rpc

[ghstack-poisoned]

gcramer23 committed:
Update on "test_combining_dpp_with_rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "test_combining_ddp_with_rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "test_combining_ddp_with_rpc"




[ghstack-poisoned]

pritamdamania87 commented:
What is the purpose of this new test? This feature is already tested here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/ddp_under_dist_autograd_test.py.

mrshenli commented on torch/testing/_internal/distributed/rpc/combining_ddp_with_rpc_test.py:
```diff
@@ -0,0 +1,191 @@
+EMBEDDING_DIM = 16
+
+
+class HybridModel(torch.nn.Module):
```
Hey @gcramer23, there might be some misunderstanding, the two onboarding tasks mentioned in the onboarding doc are the following. Did you get this task from someone else?

1. Add distributed example for CUDA RPC (https://pytorch.org/tutorials/recipes/cuda_rpc.html) to existing PyTorch CI tests.

* https://github.com/pytorch/pytorch/issues/46321 (CUDA RPC)

2. Add a tests to test_c10d.py to explicitly cover DDP on mixed dense-sparse models. Currently we have a sparse-only module tests.

* https://github.com/pytorch/pytorch/issues/54445


The second task should only use DDP (gloo backend) for mixed dense and sparse tensors.

gcramer23 replied:
@mrshenli 
My mistake must have had a misunderstanding when looking at this issue https://github.com/pytorch/pytorch/issues/46321. I will add a distributed example for the CUDA RPC. To clarify that is the reserved task in the targeted examples?

mrshenli replied:
no worries. Yep, it's the reserved task, which links to this page: https://pytorch.org/tutorials/recipes/cuda_rpc.html

Let me add your name to it.

gcramer23 committed:
Update on "test_combining_ddp_with_rpc"

adds combining_ddp_with_rpc_test.py
adds test combining_ddp_with_rpc_test to rpc_utils.py

tested using commands below on AI Platform AWS GPUs
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_combining_ddp_with_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_combining_ddp_with_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

adds combining_ddp_with_rpc_test.py
adds test combining_ddp_with_rpc_test to rpc_utils.py

tested using commands below on AI Platform AWS GPUs
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_combining_ddp_with_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_combining_ddp_with_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

adds combining_ddp_with_rpc_test.py
adds test combining_ddp_with_rpc_test to rpc_utils.py

tested using commands below on AI Platform AWS GPUs
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_combining_ddp_with_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_combining_ddp_with_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

adds cuda_distributed_rpc_test.py
adds test CudaDistributedRPCTest to rpc_utils.py

tested using commands below on AI Platform AWS GPUs
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py
adds test CudaDistributedRPCTest to rpc_utils.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

and gpu runs using 1 gpu then 2 gpus
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

H-Huang commented on torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:
```diff
@@ -0,0 +1,95 @@
+    @dist_init(setup_rpc=False)
+    def test_cuda_distributed_rpc(self):
+
+        if not torch.cuda.is_available():
+            sys.exit(TEST_SKIPS['no_cuda'].exit_code)
+
+        if torch.cuda.is_available() and torch.cuda.device_count() < 2:
+            sys.exit(TEST_SKIPS['multi-gpu'].exit_code)
```
We can use the `skip_if_not_multigpu` and `skip_if_no_gpu` decorators from  `torch.testing._internal.common_distributed` instead

H-Huang replied:
Looks like there is something wrong with the `skip_if_not_multigpu` decorator implementation as it is currently skipping all tests. Thanks @gcramer23 for bringing this up!

gcramer23 replied:
the issue I created https://github.com/pytorch/pytorch/issues/54887

H-Huang replied:
As Shen mentioned, adding `@skip_if_lt_x_gpu(2)` and `@skip_if_no_gpu` should work instead

gcramer23 replied:
added

H-Huang commented on torch/testing/_internal/distributed/rpc_utils.py:
```diff
@@ -48,6 +48,7 @@
 from torch.testing._internal.distributed.rpc.examples.reinforcement_learning_rpc_test import (
     ReinforcementLearningRpcTest,
 )
+from torch.testing._internal.distributed.rpc.cuda_distributed_rpc_test import CudaDistributedRPCTest
```
Can we put this file under `torch.testing._internal.distributed.rpc.examples` folder instead?

gcramer23 replied:
okay moved it.

H-Huang commented on torch/testing/_internal/distributed/rpc/cuda_distributed_rpc_test.py:
```diff
@@ -0,0 +1,95 @@
+import os
```
Add comment with link to example on the site

gcramer23 replied:
okay added comment with example link.

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

H-Huang reviewed (approved):
Also use `ghimport -s https://github.com/pytorch/pytorch/pull/54678` on your devserver to import this PR to phabricator.

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs


[ghstack-poisoned]

mrshenli reviewed (commented):
Is it possible to add some `self.assert*` to make sure we have parity between CPU, CUDA, and local fwd/bwd?

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

Differential Revision: [D27420409](https://our.internmc.facebook.com/intern/diff/D27420409)

[ghstack-poisoned]

gcramer23 committed:
Update on "add distributed example for CUDA RPC"

work for issue #46321

adds cuda_distributed_rpc_test.py

tested using commands below on AI Platform AWS GPUs

cpurun  pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
cpurun  pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

gpu runs using 1 gpu per node and 2 gpus per node
gpurun pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_cuda_distributed_rpc -vs
gpurun pytest test/distributed/rpc/test_process_group_agent.py -k test_cuda_distributed_rpc -vs

Differential Revision: [D27420409](https://our.internmc.facebook.com/intern/diff/D27420409)

[ghstack-poisoned]

mrshenli commented:
> do you mean self.assert on the cpu and cuda loss and gradients?

Hey @gcramer23, yep, I was referring to that. Could you please confirm after you set random seeds, model weights and input tensors have same values between CPU and CUDA runs? If so, it might relates to this issue: https://github.com/pytorch/pytorch/issues/54771

Let's hold on a bit and try again when that issue is fixed.

gcramer23 commented:
>  Could you please confirm after you set random seeds, model weights and input tensors have same values between CPU and CUDA runs?

yes the model weights and input tensors have the same values between CPU and CUDA runs.
